astropy/visualization/wcsaxes/formatter_locator.py

```diff
@@ -270,54 +306,92 @@ def locator(self, value_min, value_max):
 
-    def formatter(self, values, spacing):
+    def formatter(self, values, spacing, format='auto'):
 
         if not isinstance(values, u.Quantity) and values is not None:
             raise TypeError("values should be a Quantities array")
 
         if len(values) > 0:
+            decimal = self.decimal
+            unit = self._format_unit
+
             if self.format is None:
-                spacing = spacing.to_value(u.arcsec)
-                if spacing > 3600:
-                    fields = 1
-                    precision = 0
-                elif spacing > 60:
-                    fields = 2
-                    precision = 0
-                elif spacing > 1:
-                    fields = 3
-                    precision = 0
+                if decimal:
+                    # Here we assume the spacing can be arbitrary, so for example
+                    # 1.000223 degrees, in which case we don't want to have a
+                    # format that rounds to degrees. So we find the number of
+                    # decimal places we get from representing the spacing as a
+                    # string in the desired units. The easiest way to find
+                    # the smallest number of decimal places required is to
+                    # format the number as a decimal float and strip any zeros
+                    # from the end. We do this rather than just trusting e.g.
+                    # str() because str(15.) == 15.0. We format using 10 decimal
+                    # places by default before stripping the zeros since this
+                    # corresponds to a resolution of less than a microarcecond,
+                    # which should be sufficient.
+                    spacing = spacing.to_value(unit)
+                    fields = 0
+                    precision = len("{0:.10f}".format(spacing).replace('0', ' ').strip().split('.', 1)[1])
                 else:
-                    fields = 3
-                    precision = -int(np.floor(np.log10(spacing)))
-                decimal = False
-                unit = u.degree
+                    spacing = spacing.to_value(unit / 3600)
+                    if spacing >= 3600:
+                        fields = 1
+                        precision = 0
+                    elif spacing >= 60:
+                        fields = 2
+                        precision = 0
+                    elif spacing >= 1:
+                        fields = 3
+                        precision = 0
+                    else:
+                        fields = 3
+                        precision = -int(np.floor(np.log10(spacing)))
             else:
                 fields = self._fields
                 precision = self._precision
-                decimal = self._decimal
-                unit = self._unit
 
             if decimal:
                 sep = None
+                fmt = None
+
+                decimal = False
+                sep = 'fromunit'
```

Comment:
@Cadair - can you explain this change? Why set decimal to False here and then change sep when it was set to None on line 354?